File: src/main/kotlin/dev/jefy/connectpro/portfolio/domain/repository/JobPostRepository.java
package dev.jefy.connectpro.portfolio.domain.repository;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.List;

import dev.jefy.connectpro.management.domain.vo.CategoryId;
import dev.jefy.connectpro.portfolio.domain.model.JobPost;
import dev.jefy.connectpro.portfolio.domain.vo.JobPostId;
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioId;
import dev.jefy.connectpro.shared.infrastructure.ddd.AggregateRepository;

/**
 * @author Jôph Yamba
 */
@Repository
public interface JobPostRepository extends AggregateRepository<JobPost, JobPostId> {
    @Query("""
        select count(jobPost) > 0 from JobPost jobPost
        where jobPost.portfolioId = :portfolioId
        and jobPost.title = :title
    """)
    boolean isTitleConflict(@Param("id") PortfolioId portfolioId, @Param("title") String title);

    boolean existsByCategoryId(CategoryId categoryId);

    List<JobPost> findAllByPortfolioId(PortfolioId id);
    @Query("""
        select jobPost from JobPost jobPost
        where (:categoryId is null or jobPost.categoryId = :categoryId)
          and (:search is null or (
                lower(jobPost.title) like lower(concat('%', :search, '%'))
             or lower(jobPost.description) like lower(concat('%', :search, '%'))
        ))
    """)
    Page<JobPost> filter(@Param("search") String search, @Param("categoryId") CategoryId categoryId, Pageable page);
}

File: src/main/kotlin/dev/jefy/connectpro/portfolio/domain/repository/JobPostRepository.kt
package dev.jefy.connectpro.portfolio.domain.repository

import dev.jefy.connectpro.management.domain.vo.CategoryId
import dev.jefy.connectpro.portfolio.domain.model.JobPost
import dev.jefy.connectpro.portfolio.domain.vo.JobPostId
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioId
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface JobPostRepository : JpaRepository<JobPost, JobPostId> {

    @Query(
        """
        select case when count(jobPost) > 0 then true else false end
        from JobPost jobPost
        where jobPost.portfolioId = :portfolioId
        and jobPost.name = :title
    """
    )
    fun existsByTitleConflict
        (@Param("portfolioId") portfolioId: PortfolioId, 
         @Param("title") title: String
    ): Boolean
    
    @Query(
        """
        select case when count(jobPost) > 0 then true else false end
        from JobPost jobPost
        where jobPost.portfolioId = :portfolioId
        and jobPost.name = :title
        and jobPost.id <> :jobPostId
    """
    )
    fun existsByTitleConflictForId(
        @Param("portfolioId") portfolioId: PortfolioId, 
        @Param("title") title: String,
        @Param("jobPostId") jobPostId: JobPostId
    ): Boolean

    fun existsByCategoryId(categoryId: CategoryId): Boolean

    fun findAllByPortfolioId(id: PortfolioId): List<JobPost>

    @Query(
        """
        select jobPost from JobPost jobPost
        where (:categoryId is null or jobPost.categoryId = :categoryId)
          and (:search is null or (
                lower(jobPost.name) like lower(concat('%', :search, '%'))
             or lower(jobPost.description) like lower(concat('%', :search, '%'))
        ))
    """
    )
    fun filter(
        @Param("search") search: String?,
        @Param("categoryId") categoryId: CategoryId?,
        page: Pageable
    ): Page<JobPost>
}